rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,156 @@
+# oneDNN binary size reduction through build system.
+
+## Problem Statement.
+Several oneDNN users due to their usage model specifics would love to see their
+application linked statically with oneDNN (or shared object of the library
+itself) doesn't exceed specific disk space quota. One of such oneDNN users is
+OpenVINO toolkit. Our experiments showed that binary size of statically linked
+end-user application may take up to 64 MB [1]. Further analysis revealed several
+spots that contribute most:
+
+* The implementation of threading layer through templates. This is TBB specific,
+  OMP doesn't suffer from it. Takes ~14 MB.
+* The implementation of reorders through templates of plain C code and their
+  multiple instantiations. Takes ~9 MB.
+* The implementation of GEMM autogenerated JIT-ted kernels. Takes ~14 MB.
+* The implementation of convolution kernels apart of GEMM. Takes ~8 MB.
+* Rest functionality takes ~12 MB + ~2 MB of core API.
+
+## Background
+The reason why static linking can't solve the problem itself is the way library
+API is implemented. The call to `dnnl_primitive_desc_create` (which is also used
+underneath correspondent C++ API) relies on statically defined implementation
+lists which contain all primitives at ones. That means that if user calls the
+function (which is an essential part of the API), they get the full library in
+their application. And there are not many choices how to deal with such problem.
+One is to switch off implementations in compile time (this RFC proposal) or
+using a different API which would separate things on different levels (at least
+on primitive one).
+
+## Proposal
+To control the size, the proposal is to utilize the power of build system and
+let the user specify set of desired options. The idea is to have two sets of
+macros, one is external and would be defined by the build system based on user
+choices. This external set would control another set - internal - which lives
+inside the library and modifies implementation lists. Each primitive would get
+several macros to switch on/off certain implementations in compile time. Further
+optimizations of disk space will be provided by linker which shouldn't include
+unused code inside the end-user application unless certain scenario happens [2].
+
+As a part of this proposal, there are only two possible options to specify right
+now:
+* Propagation kind: inference or training (the default). Usage example, which
+  enables inference only:
+  - -DNNL_ENABLE_PROP_KIND=INFERENCE
```

Comment:
It would help if everything is implemented properly.
* There are some primitives that don't have backward prop kind, yes, but what does it change?
* We have a separate item to investigate how templates influence the final size.
* It uses only bwd_d. So if implemented correctly, there still be a good gain.

Reply:
I actually have the same question regarding backward path. If I need only forward + bwd_d for Convolution how I will be able to pass this configuration into cmake?

Reply:
@dmitry-gorokhov, "bwd_d for Convolution" is an essential requirement for forward de-convolution. This knowledge will live in a build system and will handle it for user as long as de-convolution is requested. This rule would be followed for any kind of such dependencies. Does it sound good?

Reply:
Yes, Seems like good enough solution